### PR TITLE
ARROW-10781 [Rust] [DataFusion] add the 'Statistics' interface in dat…

### PR DESCRIPTION
**Pull request by XiaokunDing**
Add the `Statistics` interface in data source

### CODE REVIEW

File: rust/datafusion/src/datasource/csv.rs [hunk 1]
```diff
@@ -104,4 +105,8 @@ impl TableProvider for CsvFile {
             batch_size,
         )?))
     }
+
+    fn statistics(&self) -> Option<Statistics> {
+        unimplemented!()
```
Comment (andygrove) [inline rust/datafusion/src/datasource/csv.rs:110]: I think we should return `None` here to indicate that statistics are not available (for all the current data sources).

Comment (XiaokunDing) [reply rust/datafusion/src/datasource/csv.rs:110]: OK ， Thanks andy, I'll fix it

Comment (seddonm1) [reply rust/datafusion/src/datasource/csv.rs]: I thought that you would add statistics to the structs:

```rust
/// Table-based representation of a `ParquetFile`.
pub struct ParquetTable {
    path: String,
    schema: SchemaRef,
    statistics: Option<Statistics>,
}
```

and the `statistics` method would just return the `Option<Statistics>`:

```rust
    fn statistics(&self) -> Option<Statistics> {
        self.statistics.clone()
    }
```

and `statistics` would initially be instantiated as `None` until we implement it for each source:

```rust
        Ok(Self {
            path: path.to_string(),
            schema,
            statistics: None,
        })
```

I am new to rust but this is what you were thinking @andygrove ?

File: rust/datafusion/src/datasource/datasource.rs [hunk 2]
```diff
@@ -24,6 +24,24 @@ use crate::arrow::datatypes::SchemaRef;
 use crate::error::Result;
 use crate::physical_plan::ExecutionPlan;
 
+/// The table statistics
+pub struct Statistics {
```
Comment (seddonm1) [inline rust/datafusion/src/datasource/datasource.rs:28]: The Parquet crate calls these variables. I think consistency is good so maybe they should be renamed?
    num_rows: i64,
    total_byte_size: i64,

Comment (XiaokunDing) [reply rust/datafusion/src/datasource/datasource.rs:29]: thanks Mike  I am a new to ruster too, I was fixed the code address comments by you.

File: rust/datafusion/src/datasource/csv.rs (hunk 1 above)
Comment (XiaokunDing) [reply rust/datafusion/src/datasource/csv.rs]: Ok, I fix it.

File: rust/datafusion/src/datasource/datasource.rs (hunk 2 above)
Comment (XiaokunDing) [reply rust/datafusion/src/datasource/datasource.rs:29]: OK, but I think this maybe num_rows: i64,
total_byte_size: i64,

Comment (seddonm1) [reply rust/datafusion/src/datasource/datasource.rs:29]: that code is the definition from the Parquet statistics so you should use i64

Comment (XiaokunDing) [reply rust/datafusion/src/datasource/datasource.rs:29]: OK I  change the `usize` to `i64`

Comment (seddonm1) [reply rust/datafusion/src/datasource/datasource.rs:29]: me too

**Comment by XiaokunDing** [comment]
Hi @seddonm1 @andygrove  I recreate the PR in #8866 for cherry-pick branch and by your comments,so close  this now

**Pull request closed without merging**